grudge/op.py

```diff
@@ -819,6 +820,96 @@ def _apply_inverse_mass_operator(
     return DOFArray(actx, data=tuple(group_data))
 
 
+def _apply_inverse_mass_operator_quad(
+        dcoll: DiscretizationCollection, dd_out, dd_in, vec):
+    if not isinstance(vec, DOFArray):
+        return map_array_container(
+            partial(_apply_inverse_mass_operator_quad, dcoll, dd_out, dd_in), vec
+        )
+
+    from grudge.geometry import area_element
+
+    if dd_out != dd_in:
+        raise ValueError(
+            "Cannot compute inverse of a mass matrix mapping "
+            "between different element groups; inverse is not "
+            "guaranteed to be well-defined"
+        )
+
+    actx = vec.array_context
+    dd_quad = dd_in.with_discr_tag(DISCR_TAG_QUAD)
+    dd_base = dd_quad.with_discr_tag(DISCR_TAG_BASE)
+    discr_quad = dcoll.discr_from_dd(dd_quad)
+    discr_base = dcoll.discr_from_dd(dd_base)
+
+    # Based on https://arxiv.org/pdf/1608.03836.pdf
+    # true_Minv ~ ref_Minv * ref_M * (1/jac_det) * ref_Minv
+    # Overintegration version of action on *vec*:
+    # true_Minv ~ ref_Minv * P(ref_M) * 1/P(Jac) * P(Minv*vec)
+    # P => projection to quadrature
+
+    # Compute 1/P(Jac)
+    inv_area_elements = 1/project(
+        dcoll, dd_base, dd_quad, area_element(
+            actx, dcoll, dd=dd_base,
+            _use_geoderiv_connection=actx.supports_nonscalar_broadcasting))
+
+    # Compute Minv*vec
+    def apply_minv_to_vec(vec, ref_inv_mass):
+        return actx.einsum(
+            "ij,ej->ei",
+            ref_inv_mass,
+            vec,
+            tagged=(FirstAxisIsElementsTag(),))
+
+    # Compute 1/J * vec
+    def apply_jinv_to_vec(jac_inv, vec):
+        return actx.einsum(
+            "ei,ej->ei",
```

Comment:
Do you mean just `*` here, i.e. element-wise multiplication?  This einsum will sum over each element!